RSP DropZone docs (#5076)
adobe/react-spectrum: packages/@react-spectrum/dropzone/docs/DropZone.mdx

the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License. */}

import {Layout} from '@react-spectrum/docs';
export default Layout;

import docs from 'docs:@react-spectrum/dropzone';
import {HeaderInfo, PropTable, PageDescription} from '@react-spectrum/docs';
import packageData from '@react-spectrum/dropzone/package.json';
import ChevronRight from '@spectrum-icons/workflow/ChevronRight';

```jsx import
import {DropZone} from '@react-spectrum/dropzone';
import {Heading} from '@react-spectrum/text';
import {Content} from '@react-spectrum/view';
import {IllustratedMessage} from '@react-spectrum/illustratedmessage';
import {Text} from 'react-aria-components';
import {FileTrigger} from 'react-aria-components';
import {Button} from '@react-spectrum/button';
```

---
category: Drag and drop
---

# DropZone

<PageDescription>{docs.exports.DropZone.description}</PageDescription>

<HeaderInfo
  packageData={packageData}
  componentNames={['DropZone']}

>   since="3.0.0" />

Reviewer: I _think_ adding this will mark that DropZone was added to the monopackage since `3.0.0` which isn't what we want. I think you can omit this and it will display the proper prerelease version and then will update to the monopackage version it is released with when it exits pre-release

## Example

```tsx example
import Upload from '@spectrum-icons/illustrations/Upload';

<DropZone
  maxWidth="size-3000">
  <IllustratedMessage>
    <Upload />
    <Heading>
      <Text slot="label">
        Drag and drop your file
      </Text>
    </Heading>
  </IllustratedMessage>
</DropZone>
```

## Content

> A drop zone accepts an [IllustratedMessage](IllustratedMessage.html) as a child which is comprised of three areas:  an illustration, a title, and a body. Each of these sections can be populated by providing the following components to the IllustratedMessage as children: svg, [Heading](Heading.html) (title), and [Content](Content.html) (body). A [FileTrigger](../react-aria/FileTrigger.html) is commonly paired with a DropZone to allow a user to choose files from their device. 

Author: agreed, i think an anatomy diagram would be nice for something like this. i tried my best to make one but uhh lemme know how it is haha

```tsx example
<DropZone
  maxWidth="size-3000">
  <IllustratedMessage>
    <Upload />
    <Heading>
      <Text slot="label">
        Drag and drop here
      </Text>
    </Heading>
    <Content>
      <FileTrigger>
        <Button variant="primary">Browse</Button>
      </FileTrigger>
    </Content>
  </IllustratedMessage>
</DropZone>
```

### Accessibility

A visual label should be provided to `DropZone` using a `Text` element with a `label` slot. If it is not provided, then an `aria-label` or `aria-labelledby` prop must be passed to identify the visually hidden button to assistive technology. 

## Events

> `DropZone` supports user interactions via mouse, keyboard, and touch. You can handle all of these via the `onDrop` prop. In addition, the `onDropEnter`, `onDropMove`, and `onDropExit` events are fired as the user interacts with the dropzone. 

Reviewer: nit for clarity since the following sentence implies the interaction is a drop (perhaps update DropZone docs in react-aria for this as well): 
```suggestion
`DropZone` supports drop operations via mouse, keyboard, and touch. You can handle all of these via the `onDrop` prop. In addition, the `onDropEnter`, `onDropMove`, and `onDropExit` events are fired as the user enters and exits the drop zone during a drag operation. 
```

The following example uses an `onDrop` handler to update the filled status stored in React state.

```tsx example
function Example() {
  let [isFilled, setIsFilled] = React.useState(false);

  return (
    <>
      <Draggable />
      <DropZone
        maxWidth="size-3000"
        isFilled={isFilled}
        onDrop={() => setIsFilled(true)}>
        <IllustratedMessage>
          <Upload />
          <Heading>
            <Text slot="label">
              Drag and drop here
            </Text>
          </Heading>
        </IllustratedMessage>
      </DropZone>
    </>
  );
}

> ```

Reviewer: IMO this example might be hard for the user to understand since it requires a full drop operation + a hover/keyboard drop focus to see the results of the `onDrop`. Perhaps you could mimic one of your story examples that print out the filename/text of the dropped item upon drop instead?

The `Draggable` component used above is defined below. See [useDrag](../react-aria/useDrag.html) for more details and documentation.

<details>
  <summary style={{fontWeight: 'bold'}}><ChevronRight size="S" /> Show code</summary>

```tsx example render=false export=true
import {useDrag} from '@react-aria/dnd';

function Draggable() {
  let {dragProps, isDragging} = useDrag({
    getItems() {
      return [{
        'text/plain': 'hello world',
        'my-app-custom-type': JSON.stringify({message: 'hello world'})
      }];
    }
  });

  return (
    <div {...dragProps} role="button" tabIndex={0} className={`draggable ${isDragging ? 'dragging' : ''}`}>
      Drag me
    </div>
  );
}
```
</details>

<details>
  <summary style={{fontWeight: 'bold'}}><ChevronRight size="S" /> Show CSS</summary>

```css
.draggable {
  display: inline-block;
  vertical-align: top;
  border: 1px solid gray;
  padding: 10px;
  margin-right: 20px;
  margin-bottom: 20px;
  border-radius: 4px;
  height: fit-content;
}

.draggable.dragging {
  opacity: 0.5;
}
```

</details>

## Props

<PropTable component={docs.exports.DropZone} links={docs.links} />

## Visual options

### Filled state

> The user is responsible for both managing the filled state of a drop zone and handling the associated styling. To set the drop zone in a filled state, the user must pass the `isFilled` prop.

Reviewer: mega nit: 
```suggestion
The user is responsible for both managing the filled state of a drop zone and handling the associated styling. To set the drop zone to a filled state, the user must pass the `isFilled` prop.
```

The example below demonstrates one way of styling the filled state.

```tsx example
function Example() {
  let [filledSrc, setFilledSrc] = React.useState(null);
  let [isFilled, setIsFilled] = React.useState(false);

  return (
    <>
      <DraggableImage />
      <DropZone
        isFilled={isFilled}
        maxWidth="size-3000"
        height="size-2400"
        UNSAFE_className={isFilled && 'is-filled'}
        getDropOperation={(types) =>  (types.has('image/png') || types.has('image/jpeg')) ? 'copy' : 'cancel'}
        onDrop={async (e) => {
          e.items.find(async (item) => {
            if (item.kind === 'file') {
              if (item.type === 'image/jpeg' || item.type === 'image/png') {
                setFilledSrc(URL.createObjectURL(await item.getFile()));
                setIsFilled(true);
              }
            } else if (item.kind === 'text') {
              setFilledSrc(await item.getText('image/jpeg'));
              setIsFilled(true);

>             }

Reviewer: just curious, is this portion of the code the one being hit when using `DraggableImage`?

Author: yup that's correct...not sure if there's a better way to handle it but open to changing it

Reviewer: I think it is fine as is. If we are worried about this code block being a bit too specific to how DraggableImage is implemented, we could just rely on the user dropping their own png/jpegs and just prompt them to do so via a blurb of "Try dropping a png/jpeg into the dropzone below" in the sentence before the example IMO

  border-radius: var(--spectrum-alias-border-radius-small);
}

.is-filled {
  border-style: solid;
}

```
</details>

The `DraggableImage` component used above is defined below. See [useDrag](../react-aria/useDrag.html) for more details and documentation.

<details>
  <summary style={{fontWeight: 'bold'}}><ChevronRight size="S" /> Show code</summary>

```tsx example render=false export=true
function DraggableImage() {
  let {dragProps, isDragging} = useDrag({
    getItems() {
      return [
        {
          'image/jpeg': 'https://images.unsplash.com/photo-1528722828814-77b9b83aafb2?ixlib=rb-4.0.3&ixid=M3wxMjA3fDB8MHxzZWFyY2h8MTF8fHNwYWNlfGVufDB8fDB8fHww&auto=format&fit=crop&w=900&q=60'
        }
      ];
    }
  });

  return (
    <div
      {...dragProps}
      role="button"
      tabIndex={0}
      className={`draggable ${isDragging ? 'dragging' : ''}`} >
      <img
        width="150px"
        height="100px"
        alt="Cosmic Sky"
        src="https://images.unsplash.com/photo-1528722828814-77b9b83aafb2?ixlib=rb-4.0.3&ixid=M3wxMjA3fDB8MHxzZWFyY2h8MTF8fHNwYWNlfGVufDB8fDB8fHww&auto=format&fit=crop&w=900&q=60"/>
    </div>
  );
}
```
</details>

### Replace message

> When a drop zone is in a filled state and has an object dragged over it, a message will appear in front of the drop zone. By default, this message will say "Drop file to replace". However, users can choose to customize this message through the `replaceMessage` prop. This message should describe the interaction that will occur when the object is dropped. 

Reviewer: Might be good to mention that the message should be internationalized like we do on other component docs when discussing labeling.

```tsx example
function Example() {
  let [isFilled, setIsFilled] = React.useState(false);

  return (
    <>
      <Draggable />
      <DropZone
        isFilled={isFilled}
        maxWidth="size-3000"
        replaceMessage="This is a custom message"
        onDrop={() => setIsFilled(true)}>
        <IllustratedMessage>
          <Upload />
          <Heading>
            <Text slot="label">
              Drag and drop here
            </Text>
          </Heading>
        </IllustratedMessage>
      </DropZone>
    </>
  );
}
```

### Visual feedback

> A drop zone displays visual feedback to the user when a drag hovers over the drop target by passing the `getDropOpertation` function. If a drop target only supports data of specific types (e.g. images, videos, text, etc.), then it should implement the `getDropOperation` prop and return `cancel` for types that aren't supported. This will prevent visual feedback indicating that the drop target accepts the dragged data when this is not true. [Read more about getDropOperation.](../react-aria/useDrop.html#getdropoperation)

Reviewer: ```suggestion
A drop zone displays visual feedback to the user when a drag hovers over the drop target by passing the `getDropOperation` function. If a drop target only supports data of specific types (e.g. images, videos, text, etc.), then it should implement the `getDropOperation` prop and return `cancel` for types that aren't supported. This will prevent visual feedback indicating that the drop target accepts the dragged data when this is not true. [Read more about getDropOperation.](../react-aria/useDrop.html#getdropoperation)
```

<div className={styles.exampleImage}>
  <DropZoneAnatomy />
</div>

A drop zone accepts an [IllustratedMessage](IllustratedMessage.html) as a child which is comprised of three areas:  an illustration, a title, and a body. Each of these sections can be populated by providing the following components to the IllustratedMessage as children: a SVG, a [Heading](Heading.html) (title), and a [Content](Content.html) (body). A [FileTrigger](../react-aria/FileTrigger.html) is commonly paired with a DropZone to allow a user to choose files from their device. 

```tsx example
<DropZone
  maxWidth="size-3000">
  <IllustratedMessage>
    <Upload />
    <Heading>
      <Text slot="label">
        Drag and drop here
      </Text>
    </Heading>
    <Content>
      <FileTrigger>
        <Button variant="primary">Browse</Button>
      </FileTrigger>
    </Content>
  </IllustratedMessage>
</DropZone>
```

### Accessibility

A visual label should be provided to `DropZone` using a `Text` element with a `label` slot. If it is not provided, then an `aria-label` or `aria-labelledby` prop must be passed to identify the visually hidden button to assistive technology. 

### Internationalization

> In order to internationalize a drop zone, a localized string should be passed to the `aria-label` prop.

Reviewer 2: There's a visible label as well which is an option. See https://reactspectrum.blob.core.windows.net/reactspectrum/48e8c567fffac129b7c3f07bdb43ad6b01c68dbe/docs/react-spectrum/Checkbox.html#internationalization

      Drag me
    </div>
  );
}
```
</details>

<details>
  <summary style={{fontWeight: 'bold'}}><ChevronRight size="S" /> Show CSS</summary>

```css
.draggable {
  display: inline-block;
  vertical-align: top;
  border: 1px solid gray;
  padding: 10px;
  margin-right: 20px;
  margin-bottom: 20px;
  border-radius: 4px;
  height: fit-content;
}

.draggable.dragging {
  opacity: 0.5;
}

.files{
  margin-top: 20px;
}
```

</details>

## Props

<PropTable component={docs.exports.DropZone} links={docs.links} />

## Visual options

### Filled state

The user is responsible for both managing the filled state of a drop zone and handling the associated styling. To set the drop zone to a filled state, the user must pass the `isFilled` prop.

The example below demonstrates one way of styling the filled state.

```tsx example
function Example() {
  let [filledSrc, setFilledSrc] = React.useState(null);
  let [isFilled, setIsFilled] = React.useState(false);

  return (
    <>
      <DraggableImage />
      <DropZone
        isFilled={isFilled}
        maxWidth="size-3000"
        height="size-2400"

>         UNSAFE_className={isFilled && 'is-filled'}

Reviewer 3: I don't think we want to encourage `UNSAFE_className` anywhere in the docs. Is this style needed? And if so, could we include it automatically somehow?

Author: it's not required by any means, mostly just aesthetic. i thought that it looked nicer when the border was solid versus dashed when the image filled up the dropzone area. i can try some other ways and see it goes but in the worst-case, i don't think removing this line will have any large effect

Reviewer 2: good catch, bring up in standup so we can determine if we want to build this into the component or if we need to bring it back to design
whatever the outcome, we should not use UNSAFE_className here

        getDropOperation={(types) =>  (types.has('image/png') || types.has('image/jpeg')) ? 'copy' : 'cancel'}
        onDrop={async (e) => {
          e.items.find(async (item) => {
            if (item.kind === 'file') {
              if (item.type === 'image/jpeg' || item.type === 'image/png') {
                setFilledSrc(URL.createObjectURL(await item.getFile()));
                setIsFilled(true);
              }
            } else if (item.kind === 'text') {
              setFilledSrc(await item.getText('image/jpeg'));
              setIsFilled(true);
            }
          });
        }}>
        <IllustratedMessage>
            <Upload />
            <Heading>
              <Text slot="label">
                Drag and drop photos
              </Text>
            </Heading>
        </IllustratedMessage>
        {isFilled && <img className={'images'} alt="" src={filledSrc} />}
      </DropZone>
    </>
  );
}
```

<details>
  <summary style={{fontWeight: 'bold'}}><ChevronRight size="S" /> Show CSS</summary>

```css

> .images{

Reviewer 3: style nit, I'm surprised this doesn't get formatted automatically (I guess dprint just does js(x))

```suggestion
.images {
```

  position: absolute;
  top: 0px;
  left: 0px;
  width: 100%;
  height: 100%;
  object-fit: cover;
  border-radius: var(--spectrum-alias-border-radius-small);
}

.is-filled {
  border-style: solid;
}

```
</details>

> The `DraggableImage` component used above is defined below. See [useDrag](../react-aria/useDrag.html) for more details and documentation.

Reviewer 3: I think images are draggable/droppable automatically so I think you may not need to make the `DraggableImage` component.

Author: I did try that but wasn't able to get it to work. My biggest concern was that it wouldn't be keyboard accessible which is why I decided to go with `DraggableImage` though it does add a bit of overhead to the code

Reviewer 3: Good point, sounds good!

<details>
  <summary style={{fontWeight: 'bold'}}><ChevronRight size="S" /> Show code</summary>

```tsx example render=false export=true
function DraggableImage() {
  let {dragProps, isDragging} = useDrag({
    getItems() {
      return [
        {

>           'image/jpeg': 'https://images.unsplash.com/photo-1528722828814-77b9b83aafb2?ixlib=rb-4.0.3&ixid=M3wxMjA3fDB8MHxzZWFyY2h8MTF8fHNwYWNlfGVufDB8fDB8fHww&auto=format&fit=crop&w=900&q=60'

Reviewer 3: For brevity and consistency with other docs, we may want to use imgur links. There are some already used in the repo, and I bet @snowystinger probably has more if you need them.
